### ISSUE TITLE
Unable to reproduce Dataset-1

### ISSUE THREAD

**Issue by eemario** [Cisco-Talos/binary_function_similarity#37]
Thank you for releasing the datasets!

When I use `DBs/Dataset-1/Dataset-1 creation.ipynb` to recreate Dataset-1 from `flowchart_Dataset-1.csv`, the number of selected functions differs greatly from the directly downloaded features from [Google Drive](https://drive.google.com/drive/folders/1uqZb0geb4CgDe9XEczZhNcyfBQM1TusG?usp=sharing).

E.g. The downloaded `training_Dataset-1.csv` contains 256625 functions, while the generated one contains 844841. Validation and testing sets are also very different.

Here are the steps to reproduce:

1. Loading `flowchart_Dataset-1.csv`. Now we have 8664141 functions.
2. Remove duplicated hashopcodes. 3296093 functions remain.
3. Create training (849874 functions), validation (22118 functions) and test (2424101) splits.
4. Remove common function names across splits and singleton functions. Now 844841 training, 22090 validation and 2396084 testing.

There are further steps to create positive and negative pairs (and the test set is processed with randomness), but the training set should be determined by now. Are there any other constrains to filter functions to get the released 256625 training functions (and validation and testing, too)?

P.S. I am asking about this because the reproduced testing set of Dataset-1 is so different from the released one that it significantly affects the test results.

**Comment by jimmy-sonny** [comment]
Hi @eemario, thank you for opening this issue.

`DBs/Dataset-1/training_Dataset-1.csv`, `DBs/Dataset-1/validation_Dataset-1.csv` and `DBs/Dataset-1/testing_Dataset-1.csv` are the datasets that were used to train and test the different models. Please, follow [these instructions](https://github.com/Cisco-Talos/binary_function_similarity/tree/main/DBs#download-the-features-for-each-dataset) to download them. You can also use the [`Dataset-1 sanity check.ipynb`](https://github.com/Cisco-Talos/binary_function_similarity/blob/main/DBs/Dataset-1/Dataset-1%20sanity%20check.ipynb) IPython notebook to verify that all the dataset and features are correctly downloaded.

*If the goal is to reproduce the results or do a 1-1 comparison with the trained model we released, these are the CSV files to use.*

More in detail, `training_Dataset-1.csv` contains a  subset of the functions in `features/flowchart_Dataset-1.csv` with at least 5 basic blocks. Similarly to other large projects, code and data went through several iterations, because of this and of the random sub-sampling, it's not possible to identically reconstruct it. The [`Dataset-1 creation.ipynb`](https://github.com/Cisco-Talos/binary_function_similarity/blob/main/DBs/Dataset-1/Dataset-1%20creation.ipynb) IPython notebook was cleaned up and adapted for this software release, and contains the code to generate an equivalent, but not equal (not line-by-line comparable) CSV: on the one hand the functions are filtered accurately (see the "Step 0" ... "Step 6"), and on the other hand it contains a bigger set of functions. To achieve the same dataset size it's possible to subsample the new dataset by maintaining the same distribution of functions as in `training_Dataset-1.csv`, but this has not been implemented.

*In summary, `Dataset-1 creation.ipynb` is the starting point if the goal is to create a new dataset of functions to train and test ML models, but the results cannot be compared 1-1 with what reported in the paper (for that, the provided dataset should be used instead).*
